src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java

```diff
@@ -0,0 +1,530 @@
+                                     .stream()
+                                     .filter(c -> c == indexComponent)
+                                     .map(c -> Pair.create(c, indexContext.getIndexName()))
+                                     .map(onDiskPerIndexFileMap::get)
+                                     .filter(java.util.Objects::nonNull)
+                                     .filter(File::exists)
+                                     .mapToLong(File::length)
+                                     .sum();
+        return 0;
+    }
+
+    public boolean validatePerIndexComponents(IndexContext indexContext)
+    {
+        logger.info("validatePerIndexComponents called for " + indexContext.getIndexName());
+        registerPerIndexComponents(indexContext);
+        return version.onDiskFormat().validatePerIndexComponents(this, indexContext, false);
+    }
+
+    @VisibleForTesting
+    public boolean validatePerIndexComponentsChecksum(IndexContext indexContext)
+    {
+        registerPerIndexComponents(indexContext);
+        return version.onDiskFormat().validatePerIndexComponents(this, indexContext, true);
+    }
+
+    public boolean validatePerSSTableComponents()
+    {
+        registerPerSSTableComponents();
+        return version.onDiskFormat().validatePerSSTableComponents(this, false);
+    }
+
+    public boolean validatePerSSTableComponentsChecksum()
+    {
+        registerPerSSTableComponents();
+        return version.onDiskFormat().validatePerSSTableComponents(this, true);
+    }
+
+    public void deletePerSSTableIndexComponents()
+    {
+        registerPerSSTableComponents();
+        perSSTableComponents.stream()
+                            .map(onDiskPerSSTableFileMap::remove)
+                            .filter(java.util.Objects::nonNull)
+                            .forEach(this::deleteComponent);
+        perSSTableComponents.clear();
+    }
+
+    public void deleteColumnIndex(IndexContext indexContext)
+    {
+        registerPerIndexComponents(indexContext);
+        if (perIndexComponents.containsKey(indexContext.getIndexName()))
+            perIndexComponents.remove(indexContext.getIndexName())
+                              .stream()
+                              .map(c -> Pair.create(c, indexContext.getIndexName()))
+                              .map(onDiskPerIndexFileMap::remove)
+                              .filter(java.util.Objects::nonNull)
+                              .forEach(this::deleteComponent);
+    }
+
+    public void deletePerIndexTemporaryComponents(IndexContext indexContext)
+    {
+        version.onDiskFormat()
+               .perIndexComponents(indexContext)
+               .stream()
+               .map(c -> tmpFileFor(c, indexContext))
+               .filter(File::exists)
+               .forEach(this::deleteComponent);
+    }
+
+    public void createComponentOnDisk(IndexComponent component) throws IOException
```

Comment:
nit: If we want to make the compiler warnings go away, we could throw a `@SuppressWarnings("UnstableApiUsage")` on both this and `createComponentOnDisk()`